web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx

```diff
@@ -99,11 +99,35 @@ function Information(props: { onSetUpAgentClick(): void }) {
 function AgentSetup(props: { doc: types.DocumentConnectMyComputerSetup }) {
   const ctx = useAppContext();
   const { rootClusterUri, documentsService } = useWorkspaceContext();
-  const { startAgent, markAgentAsConfigured, agentState, downloadAgent } =
-    useConnectMyComputerContext();
+  const {
+    startAgent,
+    markAgentAsConfigured,
+    isAgentConfiguredAttempt,
+    agentState,
+    downloadAgent,
+  } = useConnectMyComputerContext();
   const cluster = ctx.clustersService.findCluster(rootClusterUri);
   const nodeToken = useRef<string>();
 
+  const shouldRedirectToStatusDocument =
+    isAgentConfiguredAttempt.status === 'success' &&
+    isAgentConfiguredAttempt.data;
+
+  useEffect(() => {
+    if (shouldRedirectToStatusDocument) {
+      const statusDocument =
+        documentsService.createConnectMyComputerStatusDocument({
+          rootClusterUri,
+        });
+      documentsService.replace(props.doc.uri, statusDocument);
+    }
+  }, [
```

Comment:
A comment added. 
I also removed `useEffect` and put the code directly in the component to prevent a flash of the setup document content (I remember you used this technique some time ago).

>TBH, it would be clearer if finishing the setup explicitly called replace [...]

Definitely, but as you said it's difficult to handle both cases without making the code more complex.

Reply:
> I also removed `useEffect` and put the code directly in the component to prevent a flash of the setup document content (I remember you used this technique some time ago).

https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

This works with no issues if you're updating React state directly. With an external state like documentsService, it's worth checking if this doesn't pose any issues.

> In the above example, setSelection is called directly during a render. React will re-render the List immediately after it exits with a return statement. React has not rendered the List children or updated the DOM yet, so this lets the List children skip rendering the stale selection value.
>
> When you update a component during rendering, React throws away the returned JSX and immediately retries rendering. To avoid very slow cascading retries, React only lets you update the same component’s state during a render. If you update another component’s state during a render, you’ll see an error. A condition like items !== prevItems is necessary to avoid loops. You may adjust state like this, but any other side effects (like changing the DOM or setting timeouts) should stay in event handlers or Effects to [keep components pure.](https://react.dev/learn/keeping-components-pure)

Reply:
Yeah, I just reopened the docs and checked the console:

> Warning: Cannot update a component (`Clusters`) while rendering a different component (`DocumentConnectMyComputerSetup`). To locate the bad setState() call inside `DocumentConnectMyComputerSetup`, follow the stack trace as described in https://fb.me/setstate-in-render
>    in DocumentConnectMyComputerSetup (created by MemoizedDocument)

Reply:
Something like this? idk

https://kentcdodds.com/blog/useeffect-vs-uselayouteffect

<details>

```diff
diff --git a/web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx b/web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx
index 7bb6c78717..a4a70d09f2 100644
--- a/web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx
+++ b/web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx
@@ -14,7 +14,13 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-import React, { useCallback, useEffect, useRef, useState } from 'react';
+import React, {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'react';
 import { Box, ButtonPrimary, Flex, Text } from 'design';
 import { makeEmptyAttempt, useAsync } from 'shared/hooks/useAsync';
 import { wait } from 'shared/utils/wait';
@@ -55,16 +61,29 @@ export function DocumentConnectMyComputerSetup(
   const shouldRedirectToStatusDocument =
     isAgentConfiguredAttempt.status === 'success' &&
     isAgentConfiguredAttempt.data;
+  const isRedirectingRef = useRef(false);
 
-  // This redirect will happen when reopening the app with a status document,
-  // but also when the setup finishes.
-  if (shouldRedirectToStatusDocument) {
-    const statusDocument =
-      documentsService.createConnectMyComputerStatusDocument({
-        rootClusterUri,
-      });
-    documentsService.replace(props.doc.uri, statusDocument);
-  }
+  // useLayoutEffect instead of useEffect to prevent flashing the Setup document just before
+  // opening the Status document.
+  // TODO(ravicious): This is a hack and should be replaced with some other mechanism.
+  useLayoutEffect(() => {
+    // This redirect will happen when reopening the app with a status document,
+    // but also when the setup finishes.
+    if (shouldRedirectToStatusDocument && !isRedirectingRef.current) {
+      isRedirectingRef.current = true;
+
+      const statusDocument =
+        documentsService.createConnectMyComputerStatusDocument({
+          rootClusterUri,
+        });
+      documentsService.replace(props.doc.uri, statusDocument);
+    }
+  }, [
+    documentsService,
+    props.doc.uri,
+    rootClusterUri,
+    shouldRedirectToStatusDocument,
+  ]);
 
   return (
     <Document visible={props.visible}>

```
</details>

Reply:
Yeah, I considered `useLayoutEffect`, but then thought that I could do the redirect in the render function. 
Good to know that changing state when rendering works only for the local state. 

I applied your patch and it works fine.